### DISCUSSION TITLE
Spindle at Speed issues (Cont)

### DISCUSSION THREAD

**Discussion by rvalotta** [grblHAL/plugins#3] [Q&A] [answered]
Thought I would share the console output in hopes it might shed some light on what I'm seeing (Spindle at speed seems to alarm out after what ever delay I've set.

From what I'm guessing FS:.... is the spindle (Not sure what the first number is but second number appears to be set speed then read speed)

What I've noticed
- When I start the spindle it is showing the speed I want and doesn't seem to be picking up the spindle speed from the VFD
- It alarms 14 and now the speed is showing in the reported speed
- When it alarms its switching the spindle to off in iosender but the spindle is still going
- If i restart the spindle to the same speed the spindle seems to be happy until I change the speed

Hope this helps shed some light as to what may be going on (Be is something I've done or code)

```
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
M3S5000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
ALARM:14
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,1800>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
M3S5000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
M5
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,3883|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,2129|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,696>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
M3S5000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0>
ALARM:14
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000>
M3S5000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
M3S6000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,5000>
ALARM:14
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,1800>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
M5
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000>
M3S6000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,6000|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,6000,6000>
M5
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,6000|Ov:100,100,100|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,4645|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,3049>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,1456>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
```

**Comment by rvalotta** [comment]
Something else I just noticed... With spindle at speed disabled (value 0) the spindle is reporting its speed.

```
M3S5000
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,0|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,222>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,2046>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,3966>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000|Ov:100,100,100|A:S>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,5000,5000>
M5
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,5000|A:>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,4361|WCO:0.000,0.000,0.000>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,2767>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,1175|Ov:100,100,100>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:-5.000,-5.000,-5.000|Bf:35,1024|FS:0,0,0|WCO:0.000,0.000,0.000>
```

**Reply by terjeio** [reply]
There is something weird going on while debugging this with the STM32F4xx driver - a `delay` variable seems to somehow get a completely wrong type.
Can you try to rename it to `on_delay` and see if this changes anything?

(https://github.com/grblHAL/core/blob/0aeb53883b8dba9a62ecaa919b3609e39162245a/spindle_control.c#L87-L96)

->

```
            float on_delay = 0.0f;
            while(!(at_speed = hal.spindle.get_state().at_speed)) {
                delay_sec(0.2f, DelayMode_Dwell);
                on_delay += 0.2f;
                if(ABORTED)
                    break;
                if(on_delay >= settings.safety_door.spindle_on_delay) {
                    system_raise_alarm(Alarm_Spindle);
                    break;
                }
            }
```

**Reply by rvalotta** [reply]
Glad to try it out. Will be tomorrow before I can test it on the CNC.  Will report back.

**Comment by rvalotta** [comment]
Saw you pushed some changes so I pulled and tried again.

No go same behaviour

first part is with spindle at speed enabled (not updating the rpm) then with it disabled (rpm is getting updated)

For what its worth it is stopping the spindle now.

[VER:1.1f.20220109:]
[OPT:VNMH0SL+2,35,1024,3,0]
[NEWOPT:ENUMS,RT+,HOME,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:STM32F411]
[DRIVER VERSION:220105]
[BOARD:BlackPill]
[PLUGIN:MODBUS v0.09]
[PLUGIN:HUANYANG VFD v0.07]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:0.000,0.000,0.000]
[G55:0.000,0.000,0.000]
[G56:0.000,0.000,0.000]
[G57:0.000,0.000,0.000]
[G58:0.000,0.000,0.000]
[G59:0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000]
[G59.3:0.000,0.000,0.000]
[G28:0.000,0.000,0.000]
[G30:0.000,0.000,0.000]
[G92:0.000,0.000,0.000]
[HOME:0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000]
[PRB:0.000,0.000,0.000:0]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
M3S3000
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0|WCO:0.000,0.000,0.000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0>
ALARM:14
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|A:>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,1800>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,454>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|WCO:0.000,0.000,0.000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
[SETTINGDESCR:340|Spindle at speed tolerance as percentage deviation from programmed speed, set to 0 to disable.\nIf not within tolerance when checked after spindle on delay ($392) alarm 14 is raised.]
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
M3S3000
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,0|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,1278>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,3000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,3000|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,3000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,3000,3000|WCO:0.000,0.000,0.000>
M5
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,3000|Ov:100,100,100|A:>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,1805>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,214|WCO:0.000,0.000,0.000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|WCO:0.000,0.000,0.000>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0|Ov:100,100,100>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1024|FS:0,0,0>

**Reply by terjeio** [reply]
I have just committed another attempt at fixing this, in some configurations the VFD response was lost when in the wait for at speed delay loop.

**Comment by rvalotta** [comment, marked as answer]
That has it working thank you :)

The first

Had to make the following changes to get it to compile with STMIDE. The first one the compiler thought braces were missing based on how it was indented. The second one ifdef i2c_enable resolves true even when i2c_enable is set to 0

diff --git a/Src/driver.c b/Src/driver.c
index 8af4b52..00fa0e3 100644
--- a/Src/driver.c
+++ b/Src/driver.c
@@ -1879,8 +1879,9 @@ static bool driver_setup (settings_t *settings)

     hal.settings_changed(settings);

-    if(hal.spindle.set_state)
+    if(hal.spindle.set_state) {
        hal.spindle.set_state((spindle_state_t){0}, 0.0f);
        hal.coolant.set_state((coolant_state_t){0});
+    }


@@ -2007,7 +2008,7 @@ bool driver_init (void)
     stream_connect(serialInit(BAUD_RATE));
 #endif

-#ifdef I2C_ENABLE
+#if I2C_ENABLE
     i2c_init();
 #endif

**Reply by terjeio** [reply]
Thanks for reporting, too many changes yesterday...

I have just committed a fix.
